pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,601 @@
+        /// The parameters must be valid, all parts summing up to one whole (100%), otherwise the call will fail.
+        ///
+        /// Must be called by `root` origin.
+        ///
+        /// Purpose of the call is testing & handling unforseen circumstances.
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::force_set_inflation_params())]
+        pub fn force_set_inflation_params(
+            origin: OriginFor<T>,
+            params: InflationParameters,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            ensure!(params.is_valid(), Error::<T>::InvalidInflationParameters);
+            InflationParams::<T>::put(params);
+
+            Self::deposit_event(Event::<T>::InflationParametersForceChanged);
+
+            Ok(().into())
+        }
+
+        /// Used to force-set the inflation configuration.
+        /// The parameters aren't checked for validity, since essentially anything can be valid.
+        ///
+        /// Must be called by `root` origin.
+        ///
+        /// Purpose of the call is testing & handling unforseen circumstances.
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::force_set_inflation_config())]
+        pub fn force_set_inflation_config(
+            origin: OriginFor<T>,
+            config: InflationConfiguration,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            ActiveInflationConfig::<T>::put(config.clone());
+
+            Self::deposit_event(Event::<T>::InflationConfigurationForceChanged { config });
+
+            Ok(().into())
+        }
+
+        /// Used to force inflation recalculation.
+        /// This is done in the same way as it would be done in an appropriate block, but this call forces it.
+        ///
+        /// Must be called by `root` origin.
+        ///
+        /// Purpose of the call is testing & handling unforseen circumstances.
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::force_inflation_recalculation())]
+        pub fn force_inflation_recalculation(origin: OriginFor<T>) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let (max_emission, config) =
+                Self::recalculate_inflation(frame_system::Pallet::<T>::block_number());
+
+            ActiveInflationConfig::<T>::put(config.clone());
+
+            SafetyInflationTracker::<T>::mutate(|tracker| {
+                tracker.cap.saturating_accrue(max_emission);
+            });
+
+            Self::deposit_event(Event::<T>::ForcedInflationRecalculation { config });
+
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Used to check if inflation recalculation is supposed to happen on the next block.
+        fn is_recalculation_in_next_block(
+            now: BlockNumber,
+            config: &InflationConfiguration,
+        ) -> bool {
+            config.recalculation_block.saturating_sub(now) <= 1
```

Comment:
Actually no, because then you have a situation when recalculation can _never_ happen.

This way, even if recalculation block is set to some past block, this will trigger it.
This approach is more useful in dApp staking where there's maintenance mode but I've decided to also integrate it here.